File: src/javascript/algorithms/linkedlist/9.partitionList/partitionList.js
class Node {
  constructor(value) {
    this.value = value;
    this.next = null;
  }
}

class LinkedList {
  constructor(value) {
    this.head = null;
    this.tail = null;
    this.length = 0;
  }

  push(value) {
    const newNode = new Node(value);
    if (!this.head) {
      this.head = newNode;
      this.tail = newNode;
    } else {
      this.tail.next = newNode;
      this.tail = newNode;
    }
    this.length++;
    return this;
  }

  partitionList(x) {
    if (!this.head) return;

    let leftDummy = new Node(0);
    let rightDummy = new Node(0);
    let leftTail = leftDummy;
    let rightTail = rightDummy;

    let current = this.head;
    while (current !== null) {
      if (current.value < x) {
        leftTail.next = current;
        leftTail = current;
      } else {
        rightTail.next = current;
        rightTail = current;
      }
      current = current.next;
    }
    leftTail.next = rightDummy.next;
    rightTail.next = null;
    this.head = leftDummy.next;
  }

  printList() {
    let current = this.head;
    let output = '';
    while(current !== null) {
        output += ''+ current.value;
        current = current.next;
        if (current !== null) output += ' --> ';
    }
    console.log(output);
  }
}

const myLinkedList1 = new LinkedList();
myLinkedList1.push(7);
myLinkedList1.push(3);
myLinkedList1.push(6);
myLinkedList1.push(3);
myLinkedList1.push(5);
myLinkedList1.push(4);
myLinkedList1.push(8);

myLinkedList1.partitionList(5);
myLinkedList1.printList();

const myLinkedList2 = new LinkedList();
myLinkedList2.push(3);
myLinkedList2.push(2);

myLinkedList2.partitionList(3);
myLinkedList2.printList();

File: src/java1/algorithms/linkedlist/partitionList/PartitionList.java
package java1.algorithms.linkedlist.partitionList;

import java1.algorithms.linkedlist.Node;

public class PartitionList {

    private static Node partitionList(Node head, int x){
        Node leftDummy = new Node(0);
        Node rightDummy = new Node(0);
        Node leftTrail = leftDummy;
        Node rightTrail = rightDummy;

        Node current = head;

        while (current != null) {
            if(current.value < x){
                leftTrail.next = current;
                leftTrail = current;
            } else {
                rightTrail.next = current;
                rightTrail = current;
            }
            current = current.next;
        }

        leftTrail.next = rightDummy.next;
        rightTrail.next = null;

        return leftDummy.next;
    }

    private static void printLinkedlist(Node head) {
        while(head != null) {
            System.out.println(head.value + " ");
            head = head.next;
        }
    }

    public static void main(String[] args) {
        Node headNode1 = new Node(7);
        Node node1 = new Node(3);
        Node node2 = new Node(6);
        Node node3 = new Node(3);
        Node node4 = new Node(5);
        Node node5 = new Node(4);
        Node node6 = new Node(8);

        headNode1.setNext(node1);
        node1.setNext(node2);
        node2.setNext(node3);
        node3.setNext(node4);
        node4.setNext(node5);
        node5.setNext(node6);

        Node headNode2 = new Node(3);
        Node node7 = new Node(2);

        headNode2.setNext(node7);

        printLinkedlist(partitionList(headNode1, 5));
        printLinkedlist(partitionList(headNode2, 3));
    }
}
